content/en/blog/attribute-autocomplete-enhancement/index.md

```diff
@@ -0,0 +1,306 @@
+
+Therefore, we introduce the auto attribute correction feature that automatically corrects the attribute selection in operators whenever an attribute is deleted or renamed.
+
+### Demonstration
+
+The auto attribute correction feature updates the attribute selection in an operator under the following circumstances
+
+1. when the name of the attribute selected in the operator changes,
+2. when the attribute selected in the operator is deleted
+
+#### When attribute is renamed
+
+When user renames an attribute through an operator, the attribute name will be updated in all succeeding operators.
+
+Fig. 3 shows how the attribute selection is automatically updated in the "Filter" operator when the result attribute name is changed in its preceding "Sentiment Analysis" operator.
+
+When the user changes the name of the output attribute of the "Sentiment Analysis" operator from `"A"` to `"B"`, the `Attribute` field whose value was `"A"` in the succeeding “Filter” operator is automatically updated to `"B"`.
+
+**![](https://lh6.googleusercontent.com/Zf6wNg1uX9m_zJLkqtcKAap8kNLQttz1whOKPG1xGFolZ6it0hYMIMw2hzCLukabM0P2hoJ43M4QIjIcULYP_mAKt0N3V82L-0yuK0FOLD9xjrXSx0DSSzlqZD9Qp8QUL1KvABBpbbw5pXXrGgVpqQ)**
+
+Fig. 3, Auto correction process when attribute is renamed
+
+#### When attribute is deleted
+
+In case of an attribute selected in an operator is deleted, e.g., deselected in a projection operator, all succeeding operators containing the attribute will delete the attribute from themselves and become invalid.
+
+### Implementation
+
+To automatically correct the attribute selection in operators' properties, we need to do the following steps:
+
+1. Trigger auto correction the input attributes, i.e. input schema, changes
+2. Determine what input attribute changes (rename or delete)
+3. Correct the attribute selection
+
+The first step is done by listening to the input schema change event. We will focus on the implementation of the second and third steps.
+
+#### Determine attribute change
+
+An operator knows an attribute changes when the input schema changes. However, it can be challenging to determine what attribute is changed and what it changes to because all an operator knows about its input is the input schema, which is a list of attribute names and types.
+
+In our auto attribute correction implementation, we choose to compare the input schema before change and after change. That is, we keep a copy of the input schemas for all operators, and when an input schema changes, we compare the new input schema with the last input schema in our copy and determine the change. 
+
+When comparing the old and new input schemas, we can easily find out what attributes are unchanged, but there maybe attributes in the old and new input schemas that do not match, which means that these attributes may be 1) newly created in the new schema, 2) renamed, or 3) deleted in the new schema.
```

Comment:
a simple question here: 
assuming an old input schema `{x: str, y: int}` and a new input schema `{y: str}`, what will be the changes?

1. it could be `x renamed to y` and `deletes y`;
2. it could also be `deletes x` and `y changed to type str`.

The point is there are ambiguities. How do we support those cases? I think your later discussion regarding "edit distance" is related to this question, but I am not sure if the above case is covered.

Reply:
Yes, the edit distance algorithm selects the best one based on the distance from the ambiguities. 

> insertion and deletion to have distance 1, and update to have various distance depending on the name and type change, reflected in Table 3.

Table 3:

> | Attribute Name | Attribute Type | Distance |
> |--|--|--|
> | Unchanged | Unchanged | 0 |
> | Changed | Unchanged | 1 |
> | Unchanged | Changed | 1 |
> | Changed | Changed | 2 |

For case 1, distance = 1 (update - rename) + 1 (deletion) = 2.

For case 2, distance = 1 (deletion) + 1 (update - rename) = 2.

Since they have the same distance, we will take either result (depending on the implementation of the edit-distance library we used). This is explained in the limitation section.

> Moreover, in some cases, two possible changes may be the same edit distance. In the Table 4 and 5 example, it is possible that `Column A` is renamed to `Column C` and `Column B` is renamed to `Column D`, meanwhile it is also possible that `Column A` is renamed to `Column D` and `Column B` is renamed to `Column C`. Both possibilities have distance 2. In this case, our edit distance algorithm may take either result.
> 
> |Attribute Name|Attribute Type|
> |--|--|
> |Column A|integer|
> |Column B|integer|
> 
> Table 4, Example 2 - old input schema
> 
> |Attribute Name|Attribute Type|
> |--|--|
> |Column C|integer|
> |Column D|integer|
> 
> Table 5, Example 2 - new input schema

Reply:
Thanks for the explanation, I agree the later sections explain my question.